network/p2p/topic_validator_test.go

```diff
@@ -109,31 +107,349 @@ func TestTopicValidator(t *testing.T) {
 	err = unstakedNode.Publish(timedCtx, badTopic, data2)
 	require.NoError(t, err)
 
-	// it receives its own message
-	msg, err = unstakedSub.Next(timedCtx)
-	require.NoError(t, err)
-	require.Equal(t, msg.Data, data2)
+	// unstaked node receives its own message
+	checkReceive(timedCtx, t, data2, unstakedSub, nil, true)
 
 	// node 1 does NOT receive the message due to the topic validator
 	var wg sync.WaitGroup
-	wg.Add(1)
+
 	timedCtx, cancel := context.WithTimeout(context.Background(), time.Second)
 	defer cancel()
-	go func() {
-		msg, err = sub1.Next(timedCtx)
-		require.Error(t, err)
-		wg.Done()
-	}()
+	checkReceive(timedCtx, t, nil, sub1, &wg, false)
 
 	// node 2 also does not receive the message via gossip from the node1 (event after the 1 second hearbeat)
-	wg.Add(1)
 	timedCtx, cancel = context.WithTimeout(context.Background(), 2*time.Second)
 	defer cancel()
-	go func() {
-		msg, err = sub2.Next(timedCtx)
-		require.Error(t, err)
-		wg.Done()
-	}()
+	checkReceive(timedCtx, t, nil, sub2, &wg, false)
 
 	unittest.RequireReturnsBefore(t, wg.Wait, 5*time.Second, "could not receive message on time")
 }
+
+// TestAuthorizedSenderValidator_Authorized tests that the authorized sender validator rejects messages from nodes that are not authorized to send the message
+func TestAuthorizedSenderValidator_UnAuthorized(t *testing.T) {
+	sporkId := unittest.IdentifierFixture()
+	identity1, privateKey1 := unittest.IdentityWithNetworkingKeyFixture(unittest.WithRole(flow.RoleConsensus))
+	sn1 := createNode(t, identity1.NodeID, privateKey1, sporkId)
+
+	identity2, privateKey2 := unittest.IdentityWithNetworkingKeyFixture(unittest.WithRole(flow.RoleConsensus))
+	sn2 := createNode(t, identity2.NodeID, privateKey2, sporkId)
+
+	identity3, privateKey3 := unittest.IdentityWithNetworkingKeyFixture(unittest.WithRole(flow.RoleAccess))
+	an1 := createNode(t, identity3.NodeID, privateKey3, sporkId)
+
+	topic := engine.TopicFromChannel(engine.SyncCommittee, sporkId)
+
+	ids := flow.IdentityList{identity1, identity2, identity3}
+	translator, err := NewFixedTableIdentityTranslator(ids)
+	require.NoError(t, err)
+
+	// setup hooked logger
+	hookCalls := 0
+	hook := zerolog.HookFunc(func(e *zerolog.Event, level zerolog.Level, message string) {
+		if level == zerolog.WarnLevel {
+			hookCalls++
+		}
+	})
```

Comment:
```suggestion
	hookCalls := atomic.NewUint32(0)
	hook := zerolog.HookFunc(func(e *zerolog.Event, level zerolog.Level, message string) {
		if level == zerolog.WarnLevel {
			hookCalls.Inc()
		}
	})
```

We had [a PR](https://github.com/onflow/flow-go/pull/2293) recently that pointed out a few cases where our tests had races. I think we would be vulnerable to a race here as well, because `authorizedSenderValidator` will be called by 3 nodes concurrently. This suggestion is taken from one of the fixes they contributed.

Reply:
https://github.com/onflow/flow-go/commit/079042c5ef705456cae43752bd375f29afdec63c